### ISSUE TITLE
Offer more EC curves by default for client connections

### ISSUE THREAD

**Issue by yanavlasov** [envoyproxy/envoy#21053]
By default Envoy offers only P-256 for EC curves on client connections. This creates hard to diagnose problem when a service offers curves with longer keys only. The error message has `HANDSHAKE_FAILURE_ON_CLIENT_HELLO` and you need to use TCP dump to figure out that there is mismatch between offered EC curves on client and server.

The suggestion is to include curves with higher keys by default for client connections.

"P-256","P-384","P-521" for FIPS and "P-256","P-384","P-521","X25519" for non-FIPS builds.

**Comment by yanavlasov** [comment]
@ggreenway

**Comment by ggreenway** [comment]
I think the current setting is for the rational explained by @davidben here: https://github.com/envoyproxy/envoy/issues/10855#issuecomment-618023133.

**Comment by paul-palmer** [comment]
I presume you mean rationale and not rational. Because the argument in #10855 is not rational. It implies that the dev's analysis of the risks of using P-384 outweigh the benefits in our environment (actually every environment everywhere) when they actually know almost nothing about the risks we are balancing in our environment. This lack of P-384 support is a real pain point for us.

I can understand the Envoy team setting defaults to what they consider to be generally best practice. Even so, they should leave room for users to adjust for their specific requirements. Please add a flag feature to enable the use of P-384 and add all of these warnings about its use in the documentation.

**Comment by jdubs11** [comment]
Necro-ing these threads. NSA removed P256 from CNSA and now recommends P384 for ECDH curves. Does Envoy have plans to allow the use of P384?

**Comment by jdubs11** [comment]
@phlax can we get any traction here? I have a business need for Envoy to support P-384 curves.
Can I submit a PR?

**Comment by phlax** [comment]
> Can I submit a PR?

go for it!

cc @ggreenway

**Comment by ggreenway** [comment]
@jdubs11 what are you specifically wanting supported? The original request in this issue is to change the default curves for client connections. However, additional curves have been supported for a long time if you configure them. And your comment suggests you're just wanting support for additional curves. Example:

```
--- a/examples/tls/envoy-http-https.yaml
+++ b/examples/tls/envoy-http-https.yaml
@@ -44,3 +44,10 @@ static_resources:
       name: envoy.transport_sockets.tls
       typed_config:
         "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
+        common_tls_context:
+          tls_params:
+            ecdh_curves:
+              - X25519
+              - P-256
+              - P-384
+              - P-521
```

**Comment by ggreenway** [comment]
The default in boringssl right now is `{X25519, P-256, P-384}`: https://boringssl.googlesource.com/boringssl/+/88d7a40bd06a34da6ee0d985545755199d047258/src/ssl/extensions.cc#311

**Comment by jdubs11** [comment]
@ggreenway yes I do want support for additional curves. Can I also specify P-384 for ECDSA curves? Basically I need Envoy to support P-384 curves for key exchange and authN using ECDSA certificates. In my current config, Envoy throws an error when I try to have it present an ECDSA cert generated on a P-384 curve. 

Similarly if I start Envoy with an ECDSA cert generated on a P-256 curve, and then have a client using an ECDSA cert generated on P-384 curve try and connect using mTLS. Envoy throws an error about not supporting P-384. If I specify the curve as P-256 in the connection attempt just described, it does succeed. But P-384 is minimum for CNSA
